### ISSUE TITLE
Pianobar in Termux: /!\ Cannot open audio device

### ISSUE THREAD

**Issue by tcaddy** [PromyLOPh/pianobar#664]
### Cannot open audio device
I was able to get pianobar to compile in the [Termux](https://wiki.termux.com/wiki/Main_Page) app on a Chromebook.  However, when I try to run the app I get console errors that say:
```bash
/!\ Cannot open audio device
```

Pulseaudio is running.  There is a Termux package called `play-audio`.  It will play a sample MP3 file from the Termux command line.

### Your environment
* version of pianobar: 2018.06.22-dev
* your Linux distribution and release version: output of `termux-info` command:
```bash
bash-4.4$ termux-info
Updatable packages:
All packages up to date
System information:
Linux localhost 3.18.0-18024-gc429fbc8b604 #1 SMP PREEMPT Tue Jun 19 04:33:31 PDT 2018 i686 Android
Termux-packages arch:
i686
Android version:
7.1.1
Device manufacturer:
google
Device model:
Acer Chromebook R11 (CB5-132T / C738T)
```

* ffmpeg/libav version and the flags it was compiled with (if you compiled yourself):  ffmpeg -v 4.0.1 installed via `pkg` app in Termux
* your config file: barebones config with user and password

### Steps to reproduce
Getting pianobar to compile in the Termux environment was tricky and I didn't document it, so this is my best guess.  LMK if I missed anything.

#### install packages
Use the `pkg install` command to install:
* clang
* ffmpeg-dev
* json-c-dev
* libcurl-dev
* libgcrypt-dev
* make
* pulseaudio

#### install libao
There is no `libao` package available so I had to build it from source.
* download the source, cd into
* `./configure --prefix=$PREFIX`
* `make`
* `make install`

#### edit Makefile
On line 20 of the Make file, change the else block to return this: `CC:=gcc -std=c99`

#### make pianobar
(I had to use `make` b/c `gmake` isn't listed as a package in Termux)
* `make clean`
* `make`

#### start pulseaudio
`pulseaudio`

#### run pianobar
`pianobar`

### Expected behaviour
It should play audio after getting station and song.

### Actual behaviour
It gives the error message and skips to the next track.
```bash
/!\ Cannot open audio device
```

**Comment by tcaddy** [comment]
This is the readme for the `play-audio` application:https://github.com/termux/play-audio/blob/master/README.md

It says it is using OpenSL ES: https://www.khronos.org/opensles/

**Comment by tcaddy** [comment]
I also followed [this troubleshooting for pulseaudio](https://en.wikibooks.org/wiki/Configuring_Sound_on_Linux/Pulse_Audio/Testing), which worked for playing a .WAV file.

The output of the `list-sinks` from the `pacmd` shell might be of interest:
```
$ pacmd
Welcome to PulseAudio 12.0! Use "help" for usage information.
>>> load-sample test /data/data/com.termux/files/home/test.wav
>>> list-sinks
1 sink(s) available.
  * index: 1
        name: <OpenSL_ES_sink>
        driver: <module-sles-sink.c>
        flags: DECIBEL_VOLUME LATENCY FLAT_VOLUME DYNAMIC_LATENCY
        state: SUSPENDED
        suspend cause: IDLE
        priority: 1000
        volume: front-left: 65536 / 100% / 0.00 dB,   front-right: 65536 / 100% / 0.00 dB
                balance 0.00
        base volume: 65536 / 100% / 0.00 dB
        volume steps: 65537
        muted: no
        current latency: 0.00 ms
        max request: 344 KiB
        max rewind: 344 KiB
        monitor source: 1
        sample spec: s16le 2ch 44100Hz
        channel map: front-left,front-right
                     Stereo
        used by: 0
        linked by: 0
        configured latency: 0.00 ms; range is 0.50 .. 2000.00 ms
        module: 16
        properties:
                device.description = "OpenSL ES Output"
                device.class = "abstract"
                device.icon_name = "audio-card"
>>> play-sample test OpenSL_ES_sink
Playing on sink input #0
>>> exit
```

**Comment by PromyLOPh** [comment]
Can you recall which audio backends were enabled by libao’s ./configure?
It might be helpful to switch on libao’s debug output by adding
``debug=1`` to ``~/.libao``. You can force the pulseaudio driver by
adding ``default_driver=pulse`` to the same file.

**Comment by tcaddy** [comment]
I think the libao was configured for pulseaudio and something that might have been called "null output".

My libao conf file has a different name.  I will try to rename it and add the debug option

**Comment by tcaddy** [comment]
With the debug option, it shows this error on startup:
```
debug: Loading driver plugins from /data/data/com.termux/files/usr/lib/ao/plugins-4...
ERROR: Failed to load plugin /data/data/com.termux/files/usr/lib/ao/plugins-4/libpulse.so => dlopen() failed
```

Looks like the problem is with libao.

**Comment by edward-p** [comment]
@tcaddy I think we need port it by using another audio playback library that suitable for Android just like the project "play-audio", rather than using libao and pulse audio.

**Comment by PromyLOPh** [comment]
@tcaddy: I assume libpulse.so does not exist?
@edward-p: Writing a plugin for libao would be the better solution imo.
It’s really not that difficult and every other projects using libao
benefits.

**Comment by edward-p** [comment]
@PromyLOPh
@tcaddy

 It seems that libpulse on termux isn't fully functional now, which can be tracked here [Integrating ALSA and Pulseaudio into Termux #821](https://github.com/termux/termux-packages/issues/821)

**Comment by tcaddy** [comment]
So this runs without error, but there is no sound:
```LD_PRELOAD=$PREFIX/lib/libpulse.so pianboar```

**Comment by edward-p** [comment]
@tcaddy 
I patched `configure.ac` with `LIBAO_LA_LDFLAGS="-lpulse"` for building `libao` , it works fine without `LD_PRELOAD=$PREFIX/lib/libpulse.so`.
see here [Add packages: pianobar, libao #2641](https://github.com/termux/termux-packages/pull/2641/files)

**Comment by tcaddy** [comment]
Correct, it works for me when I manually start pulseaudio, too.

This is pretty cool.  Thanks for your help!

**Comment by edward-p** [comment]
@tcaddy
But there is an buffer underrun issue with pulseaudio which makes the sound choppy randomly.

**Comment by PromyLOPh** [comment]
Actually, this could be pianobar’s fault. If you look at the player code
you will find it decodes chunks of data, then passes them to libao and
only after libao returned, decodes the next chunk. It should do this in
parallel.

**Comment by edward-p** [comment]
@PromyLOPh 
I think you're right. After I add `usleep(20000)` after the `ao_play` function, this issue reproduced on my computer. It seems that mobile devices can't do decoding as fast as computer. We need fix it by doing decoding and playing in parallel with a buffer.

**Comment by edward-p** [comment]
@PromyLOPh I tried putting `ao_play` function in another thread which created by `BarPlayerThread` . The `BarPlayerThread` prepare frames to play and add them into a `STAILQ` list which defined in `<sys/queue.h>`. The other thread access the head of the queue each time, play it then remove it from the queue head. Although with some new issues, the sound is fine on my computer. Howerver, I still get the random choppy sounds on termux. This could be pulseaudio-on-termux's fault now.

**Comment by PromyLOPh** [comment]
Can you share your code? Does the problem exist with other programs that
use pulseaudio on termux and don’t use libao (paplay for example)?

**Comment by edward-p** [comment]
@PromyLOPh [here](https://github.com/edward-p/pianobar/commit/64b25303716b0d80839dbe41cb8dcb6dfe153eb6) is my changes. It looks not so good, I think. Because I'm not familiar with ffmpeg, I can't rewrite the whole player. I tried cmus on termux which doesn't use libao, it works perfectly.

**Comment by PromyLOPh** [comment]
Well, it’s not too bad. Copying av frames/packets is obviously not the
most efficient thing to do, but that should’t be the problem. Also
the new AoPlayThread spins until queue_element!=NULL, which can consume
a lot of CPU time. Still, in theory it should fix the issue.

A different approach would be to offload playback to an external program
entirely. I already tried that, see
https://github.com/PromyLOPh/pianobar/tree/externalplay and – except for
a few unsolved issue – it works quite well. Unfortunately that branch is
very outdated and not easily portable to master.

**Comment by edward-p** [comment]
@PromyLOPh That's an ingenious idea to use external program and pipe. I've learned a lot from you, thanks! Now I think temux users can enjoy `pianobar` on their  Android devices. :+1:

**Comment by edward-p** [comment]
@PromyLOPh I tried to avoid copying av frames/packets, and adjust the time when `AoPlayThread` being created.  When decoding is done, insert an "EOF" element to the end of the queue and wait for  `AoPlayThread` to quit (when the thread get the "EOF" element from the queue or when shouldQuit(player) becomes true). Now it works quite well on my phone. See [https://github.com/edward-p/pianobar](https://github.com/edward-p/pianobar)
Still, I have no idea how to deal with the CPU time consuming caused by the thread spinning until queue_element!=NULL.
No matter how, the choppy sound issue is already gone :).

**Comment by edward-p** [comment]
Rebased and sent pull request [pull/665](https://github.com/PromyLOPh/pianobar/pull/665/files).

**Comment by edward-p** [comment]
Another issue may lead to random choppy sound.

``` bash
$ termux-audio-info 
{
  "PROPERTY_OUTPUT_SAMPLE_RATE": "48000",
  "PROPERTY_OUTPUT_FRAMES_PER_BUFFER": "192",
  "AUDIOTRACK_SAMPLE_RATE": 48000,
  "AUDIOTRACK_BUFFER_SIZE_IN_FRAMES": 3844,
  "AUDIOTRACK_SAMPLE_RATE_LOW_LATENCY": 48000,
  "AUDIOTRACK_BUFFER_SIZE_IN_FRAMES_LOW_LATENCY": 384,
  "BLUETOOTH_A2DP_IS_ON": false,
  "WIREDHEADSET_IS_CONNECTED": true
}
```

As you can see `AUDIOTRACK_SAMPLE_RATE_LOW_LATENCY=48000`, `PROPERTY_OUTPUT_SAMPLE_RATE=48000`.

But in pulseaudio
``` bash
$ pulseaudio --dump-conf
...
enable-lfe-remixing = no
lfe-crossover-freq = 0
default-sample-format = s16le
default-sample-rate = 44100
alternate-sample-rate = 48000
default-sample-channels = 2
default-channel-map = front-left,front-right
default-fragments = 4
default-fragment-size-msec = 25
...
```

The `default-sample-rate` is set to `44100` by default after pulseaudio installed on Termux.

This cause the audio sample need to be converted to 48000 by the android system's audioserver before output, which waste a lot of CPU resources (I have monitored it by `htop`)

So it's important to modify the `$PREFIX/etc/pulse/daemon.cfg` to match the `default-sample-rate` to your device.

**Comment by tomty89** [comment]
While setting pulse sink rate to the current optimum is mostly a good thing to do, it shouldn't cause any underrun even if it is on a non-optimal rate (at least not with my recent rewrite of the sink module).

So the sound shouldn't be choppy unless your device is really low end that it literally can't afford Android's (redundant) resampling. You can report back with a pulse log (e.g. `pulseaudio -vvv --daemonize=no &> ~/pulse.log &`) anyway.

Note that 48kHz is not some universal optimum. The value depends on your (current) audio device and even your Android configuration. So there isn't really an issue in Termux/pulse on the respect.

SLES in Android provides no means for getting the current optimal rate. AAudio does though, while it is only available in Oreo or later. Might port the SLES sink module to an (extra) AAudio one when I have time.

**Comment by edward-p** [comment]
@tomty89 I think this could be the problem of LineageOS 15.1 I'm using, which is in development and unofficial build for a sdm845 device. I get the random choppy sound mostly when I'm using my phone(e.g. screen rotated).
Here is the log:
[pulse.log](https://github.com/PromyLOPh/pianobar/files/2244878/pulse.log)
 No this issue on my older phone (also LineageOS 15.1).

p.s. pianobar talks to libao instead of talking to pulseaudio directly.

**Comment by tomty89** [comment]
Actually I think it's a libao issue. Its code for negotiating buffer/latency with pulse doesn't seem to be really proper, while at the same time it requests really little of that (base on a 20ms factor). Normally the configured latency of the sink would prevent that from happening, but with libao's code weird things seem to happen.

There is a workaround, which is to set `buffer_time` in libao.conf. However libao seems to be broken in parsing its conf as well, in which the option can at max be set to 44, otherwise the value would become enormous somehow.

The reason that the underruns happen more often when a non-optimal rate is used is, Android has a higher buffer/latency requirement with such track. It also gets way higher if you were on Bluetooth audio, where 44 is hardly enough.

Didn't test with pianobar, but could reproduce similar problem with cmus if it outputs to pulse via libao. No problem if it outputs to pulse via libpulse directly.

**Comment by edward-p** [comment]
@tcaddy Yeah, I think you're right.  Setting `buffer_time` in libao.conf works. Have you changed the `AO_SYSTEM_CONFIG` to `/data/data/com.termux/files/usr/etc` which defined in [include/ao/ao_private.h](https://github.com/timonwong/libao/blob/master/include/ao/ao_private.h), line 54?

**Comment by tomty89** [comment]
Nah I was testing with cmus/libao/libpulse in an Arch proot with the help of this [trick](https://wiki.archlinux.org/index.php/PulseAudio/Examples#Allowing_multiple_users_to_use_PulseAudio_at_the_same_time). Didn't bother to build them under Termux.

This is what happen with `buffer_time=44`:
```
I: [pulseaudio] sink-input.c: Created input 0 "libao[cmus] playback stream" on OpenSL_ES_sink with sample spec s16le 2ch 48000Hz and channel map front-left,front-right
I: [pulseaudio] sink-input.c:     media.name = "libao[cmus] playback stream"
I: [pulseaudio] sink-input.c:     application.name = "libao[cmus]"
I: [pulseaudio] sink-input.c:     native-protocol.peer = "UNIX socket client"
I: [pulseaudio] sink-input.c:     native-protocol.version = "32"
I: [pulseaudio] sink-input.c:     application.process.id = "8352"
I: [pulseaudio] sink-input.c:     application.process.user = "tom"
I: [pulseaudio] sink-input.c:     application.process.host = "localhost"
I: [pulseaudio] sink-input.c:     application.process.binary = "cmus"
I: [pulseaudio] sink-input.c:     application.language = "C"
I: [pulseaudio] sink-input.c:     application.process.machine_id = "1e448bfd85d542a4a1290e5ad1baf413"
I: [pulseaudio] sink-input.c:     module-stream-restore.id = "sink-input-by-application-name:libao[cmus]"
I: [pulseaudio] protocol-native.c: Requested tlength=44.00 ms, minreq=11.00 ms
D: [pulseaudio] protocol-native.c: Adjust latency mode enabled, configuring sink latency to half of overall latency.
D: [pulseaudio] protocol-native.c: Requested latency=11.00 ms, Received latency=125.00 ms
D: [pulseaudio] memblockq.c: memblockq requested: maxlength=10560, tlength=28224, base=4, prebuf=26116, minreq=2112 maxrewind=0
D: [pulseaudio] memblockq.c: memblockq sanitized: maxlength=10560, tlength=10560, base=4, prebuf=8452, minreq=2112 maxrewind=0
I: [pulseaudio] protocol-native.c: Final latency 180.00 ms = 33.00 ms + 2*11.00 ms + 125.00 ms
```

This is what happen with `buffer_time=45`:
```
I: [pulseaudio] sink-input.c: Created input 0 "libao[cmus] playback stream" on OpenSL_ES_sink with sample spec s16le 2ch 48000Hz and channel map front-left,front-right
I: [pulseaudio] sink-input.c:     media.name = "libao[cmus] playback stream"
I: [pulseaudio] sink-input.c:     application.name = "libao[cmus]"
I: [pulseaudio] sink-input.c:     native-protocol.peer = "UNIX socket client"
I: [pulseaudio] sink-input.c:     native-protocol.version = "32"
I: [pulseaudio] sink-input.c:     application.process.id = "8393"
I: [pulseaudio] sink-input.c:     application.process.user = "tom"
I: [pulseaudio] sink-input.c:     application.process.host = "localhost"
I: [pulseaudio] sink-input.c:     application.process.binary = "cmus"
I: [pulseaudio] sink-input.c:     application.language = "C"
I: [pulseaudio] sink-input.c:     application.process.machine_id = "1e448bfd85d542a4a1290e5ad1baf413"
I: [pulseaudio] sink-input.c:     module-stream-restore.id = "sink-input-by-application-name:libao[cmus]"
I: [pulseaudio] protocol-native.c: Requested tlength=5592394.23 ms, minreq=5592394.21 ms
D: [pulseaudio] protocol-native.c: Adjust latency mode enabled, configuring sink latency to half of overall latency.
D: [pulseaudio] protocol-native.c: Requested latency=0.00 ms, Received latency=125.00 ms
D: [pulseaudio] memblockq.c: memblockq requested: maxlength=4194304, tlength=2147503376, base=4, prebuf=1073763690, minreq=1073739690 maxrewind=0
D: [pulseaudio] memblockq.c: memblockq sanitized: maxlength=4194304, tlength=4194304, base=4, prebuf=4, minreq=4194304 maxrewind=0
I: [pulseaudio] protocol-native.c: Final latency 21970.33 ms = 22347776.00 ms + 2*21845.33 ms + 125.00 ms
```

**Comment by edward-p** [comment]
Only two underruns get each song at the start with `buffer_time=44` I think this is acceptable.
```
I: [pulseaudio] sink-input.c: Created input 0 "libao[pianobar] playback stream" on OpenSL_ES_sink with sample spec s16le 2ch 44100Hz and channel map front-left,front-right
I: [pulseaudio] sink-input.c:     media.name = "libao[pianobar] playback stream"
I: [pulseaudio] sink-input.c:     application.name = "libao[pianobar]"
I: [pulseaudio] sink-input.c:     native-protocol.peer = "UNIX socket client"
I: [pulseaudio] sink-input.c:     native-protocol.version = "32"
I: [pulseaudio] sink-input.c:     application.process.id = "11359"
I: [pulseaudio] sink-input.c:     application.process.user = "u0_a118"
I: [pulseaudio] sink-input.c:     application.process.host = "localhost"
I: [pulseaudio] sink-input.c:     application.process.binary = "pianobar"
I: [pulseaudio] sink-input.c:     application.process.machine_id = "localhost"
I: [pulseaudio] sink-input.c:     module-stream-restore.id = "sink-input-by-application-name:libao[pianobar]"
I: [pulseaudio] protocol-native.c: Requested tlength=43.99 ms, minreq=11.00 ms
D: [pulseaudio] protocol-native.c: Adjust latency mode enabled, configuring sink latency to half of overall latency.
D: [pulseaudio] protocol-native.c: Requested latency=11.00 ms, Received latency=125.00 ms
D: [pulseaudio] memblockq.c: memblockq requested: maxlength=9700, tlength=25932, base=4, prebuf=23996, minreq=1940 maxrewind=0
D: [pulseaudio] memblockq.c: memblockq sanitized: maxlength=9700, tlength=9700, base=4, prebuf=7764, minreq=1940 maxrewind=0
I: [pulseaudio] protocol-native.c: Final latency 179.99 ms = 32.99 ms + 2*11.00 ms + 125.00 ms
D: [sles-sink] protocol-native.c: max_request changed, trying to update from 9700 to 25932.
D: [sles-sink] protocol-native.c: Failed to increase tlength
D: [pulseaudio] sink.c: OpenSL_ES_sink: state: IDLE -> RUNNING
D: [pulseaudio] core-subscribe.c: Dropped redundant event due to change event.
D: [sles-sink] protocol-native.c: Requesting rewind due to end of underrun.
D: [sles-sink] protocol-native.c: Requesting rewind due to end of underrun.
```
No underrun get during the song is playing. I haven't tested with Bluetooth audio yet.

**Comment by tomty89** [comment]
@edward-p Okay I found and fixed problem. Here's the patch:
```
diff --git a/src/plugins/pulse/ao_pulse.c b/src/plugins/pulse/ao_pulse.c
index 2d10d57..534af63 100644
--- a/src/plugins/pulse/ao_pulse.c
+++ b/src/plugins/pulse/ao_pulse.c
@@ -41,7 +41,7 @@
 #include <ao/ao.h>
 #include <ao/plugin.h>

-#define AO_PULSE_BUFFER_TIME 20000
+#define AO_PULSE_BUFFER_TIME 50000

 /* Unfortunately libao doesn't allow "const" for these structures... */
 static char * ao_pulse_options[] = {
@@ -256,12 +256,10 @@ int ao_plugin_open(ao_device *device, ao_sample_format *format) {
     }

     /* buffering attributes */
-    battr.prebuf = battr.minreq = battr.fragsize = -1;
+    battr.prebuf = battr.minreq = battr.fragsize = battr.maxlength = -1;

-    battr.tlength = (int)(internal->buffer_time * format->rate) / 1000000 *
-      ((format->bits+7)/8) * device->output_channels;
-    battr.minreq = battr.tlength/4;
-    battr.maxlength = battr.tlength+battr.minreq;
+    battr.tlength = internal->buffer_time * format->rate / 1000000 *
+      (format->bits / 8) * device->output_channels;

     internal->simple = pa_simple_new(internal->server, t, PA_STREAM_PLAYBACK,
                                      internal->sink, t2, &ss,
```
If you don't feel like changing the current way libao requests buffer and its default, you can use this instead, which only fixes the bug:
```
diff --git a/src/plugins/pulse/ao_pulse.c b/src/plugins/pulse/ao_pulse.c
index 2d10d57..f3e9376 100644
--- a/src/plugins/pulse/ao_pulse.c
+++ b/src/plugins/pulse/ao_pulse.c
@@ -258,8 +258,8 @@ int ao_plugin_open(ao_device *device, ao_sample_format *format) {
     /* buffering attributes */
     battr.prebuf = battr.minreq = battr.fragsize = -1;

-    battr.tlength = (int)(internal->buffer_time * format->rate) / 1000000 *
-      ((format->bits+7)/8) * device->output_channels;
+    battr.tlength = internal->buffer_time * format->rate / 1000000 *
+      (format->bits / 8) * device->output_channels;
     battr.minreq = battr.tlength/4;
     battr.maxlength = battr.tlength+battr.minreq;

```

**Comment by edward-p** [comment]
@tomty89 That's great, thanks! I will apply the first when building libao for termux.

**Comment by iamlord** [comment]
Works pretty good. Sometimes you have to force stop the APP if it doesn't recognize the device and keeps playing. Keeps running during phone calls. Great work guys! Big fan promy. Pianobar has definitely enriched my life. I read the man pages for pianobar but I do not see where I can store my creds for the mobile version. It is amazing what you guys have done I am still learning about packages on mobile.

**Comment by edward-p** [comment]
@iamlord 

Termux has its own `HOME` variable just like all Unix-Likes.
So you just create a `$HOME/.config/pianobar/config` like this:

```
password = your-password
user = yourname@example.com
```

Then `pianobar` can parse it and login automatically.

**Comment by ijustlovemath** [comment]
Sorry for the necro-bump, but I can confirm this now works with minimal configuration! I'm going to llace this here since this is the first result when you search "pianobar termux" 

Here's how I did it:

1. Install the prereq packages:

`pkg install pulseaudio pianobar`

2. Create `~/.libao` with content

`default_driver=pulse`

3. (optional) Create config file `~/.config/pianobar/config` with content

```
user = your.pianobar@user.email
password = your_plaintext_password
```
4. Run pulseaudio daemon

`pulseaudio --daemon`

5. Run pianobar

`pianobar`

I'm currently working out how to use termux-notify to get media controls for this, probably with that other pianobar fifo tool. 

---

Here's a basic script I'm using for media controls; I placed control-pianobar.sh in the same directory as pianobar's config file to make this, and renamed it pianobar.sh

```
#!/bin/bash                                                       
set -eu                                                           
pbr="env /data/data/com.termux/files/home/.config/pianobar/pianobar.sh"
termux-notification --type media --media-next "$pbr next || termux-toast next failed" --media-pause "$pbr p || termux-toast pause failed" --media-play "$pbr play || termux-toast play failed" --media-previous "termux-toast cant go previous" --title pianobar --id pianobar
```

You might have to do `termux-fix-shebang` on your scripts if you're getting "file not found" errors. I did not enable `event_command` because it has a few issues since last being written.


I've also consolidated a termux-specific set of scripts for controlling pianobar thru notifications: https://github.com/ijustlovemath/pbr
